### ISSUE TITLE
Randomize the order of sent contributions in batches

### ISSUE THREAD

**Issue by mathieui** [streetcomplete/StreetComplete#3507]
**Use case**
Disclaimer: I did not look at the code before opening this, maybe it is an irrelevant request, and I am no expert in OSM contributions.

When using StreetComplete, I do not want my movements to be tracked through the various quests I fulfill along the way.It seems that it would currently be possible to know such things (and incidentally, precisely where I live, work, shop, etc…) through my OSM profile history.

**Proposed Solution**

To add some noise into this very personal history, my proposed solution is to randomize the order in which grouped modifications are sent (assuming that is a thing), or even individual contributions, so that no clear point of origin or destination can be inferred. Obviously due to the nature of OSM mapping, we cannot change the content itself, which already reveals some information about the mapper, but this proposed feature would make it at least slightly harder.

**Comment by westnordost** [comment]
This is probably not possible. It is not possible to upload the edits in any order because some edits made with StreetComplete are dependent on another. For example you split a way, and then answered other questions on the different sections. So, a random order would not work. Right now I have no idea how a conditional random order could be achieved without badly messing up the logic. The edits do not refer to each other like "I depend on this or that edit" but know nothing of another - the dependence is implicit based on the order in the upload queue.

**Comment by westnordost** [comment]
Maaaybe it would be possible to always upload split-ways (in order) first, then randomize the rest. But I am not sure about that and I am somewhat wary to touch this central logic of the app.

**Comment by smichel17** [comment]
Perhaps something along these lines this could be cleanly integrated with the logic for deciding when to close a changeset? For example, when a changeset includes an edit which is order-dependent, then that changeset is not randomized. Onviously this wouldn't *solve* the problem, but it might be a relatively safe/clean first step.

Independently, I was going to propose for auto-upload to wait to upload until it determines that the changeset should be closed-- mostly for battery savings, partially to avoid publishing realtime location data.

**Comment by smichel17** [comment]
Also see #3208 and #3466 for recent related discussions.

**Comment by biketeur** [comment]
Eine Änderung der jetzigen Verarbeitungsweise bzw. randomisierte Speicherung von Quests bringt nicht das Gewünschte:
ein User geht an einem Tag X von A nach B und beantwortet etliche Quests. Egal in welcher Reihenfolge die Quests hochgeladen werden, wenn alle beantworteten Quests des Tages X auf einer Karte markiert werden, ist der Weg des users ersichtlich. Offen ist vielleicht, ob er von A nach B oder von B nach A gegangen ist. Aber dies kann man durch weitere Überlegungen evtl. auch herausfinden

**Comment by westnordost** [comment]
Ja, korrekt.

**Comment by timothywashere** [comment]
Are you talking about live tracking whilst you're using the app or at some time in the future?
I can see the concern for the first but surely the answer is just to turn off the auto upload option.
I don't see the concern with the second, unless you are worried someone will work out your way home from work or something?

**Comment by westnordost** [comment]
So to be clear: The app uploads edits in exactly the same order as they were created, regardless if the auto upload option is on or off.

Since the upload of the individual edits is always some milliseconds apart, it is theoretically possible to derive from the timestamps from where (start) to where (end) a survey with this app went. If the edits were instead uploaded in a random order, the path the survey went would still be deriveable but not whether the user went from A to B or from B to A.

Though, I'd say that this additional detail is somewhat negligible, considering the more fundamental privacy issue that exposes the location(s) the user went while contributing.
But this is a general issue anyway with contributing data on-site to OpenStreetMap and is a combination of the requirement that anonymous contributions are not possible in OSM and maybe also the lax protection of the data (you don't even need to be logged and thus acknowledge the terms and conditions to see who contributed what at what time).

The only workaround for a contributor to improve privacy for himself here is to use a range of several accounts. E.g. semi-anonymous secondary OSM accounts that are not tied to your usual OSM activities or identity.

**Comment by mnalis** [comment]
I agree, the path taken by user is trivially reconstructed regardless of POI order (as users moves using their own foot or some vehicle, and not using teleport), and thus doing that would only produce _false sense of privacy_, which is arguably actually **much worse** than explicitly stating what the privacy issues are (as SC currently does). 

As noted, helping improve OSM on the ground (by using apps like SC) is necessarily going to lead to revealing where one has been.

@mathieui There are few things one might employ if they were so inclined to make it somewhat harder for others to mine publicly available OSM data to undermine their privacy (but at a loss of some convenience, as always):

- do not enter any user identifiable information into OSM (including your known e-mail address or recognizable username), so even when bad actors find out where someone have been, they will not know it is where **you** have been.
- use those OSM account(s) only for SC mapping (i.e. no setting home, no setting friends, no OSM messages etc).
- avoid mapping several kilometers around location-sensitive places (like home or work - that probably means somewhat less contributions though); only start mapping when you are far away from them (but also do not be too obvious - if you've mapped most of the city except two kilometer-wide circles, such geoblocking is as telling as if your were mapping them daily! see http://hdyc.neis-one.org/ / https://yosmhm.neis-one.org/ for example)
- never use auto-upload functionality of SC (default!); instead, accumulate several weeks (or even months) of updates and upload them manually all at once (that will likely lead to some lost work if the area is frequently mapped by others, though) - preferably some time after they have been mapped and not immediately afterwards
- do not map frequently used routes (like daily going to work or to home); instead, concentrate your efforts on mapping when visiting public and rarely used ones (like weekend trips to different locations; avoid mapping when visiting friends/family though)
- use different OSM accounts to map different parts of city (do not confuse them, though. Probably ruins gamification effect of SC too; although I guess one can practice competing their own alt-personas one against another :smile:)
- use modified fork of StreetComplete, which does not set `source=survey`, fakes the editor name to simulate JOSM or iD, and uploads all changes in one changeset

(Of course, none of that would help at all if one was explicitly tracked outside of OSM, like by some bribed google employee having access to location data helpfully submitted by Android automatically, or _$DEITY_ forbid if one was person of interest of some shady organizations or governments)

**Comment by matkoniecz** [comment]
If someone really, really, really needs to keep locations fully private I would rather advise to not map by survey.

Even after applying solutions above, given low density of OSM mappers, in most cases it would be possible to reconstruct it.

**Comment by smichel17** [comment]
Or invite all your friends/neighbors/random people on the street to do 
mapping, to increase the density :D

**Comment by tiptoptom** [comment]
> * never use auto-upload functionality of SC (default!); instead, accumulate several weeks (or even months) of updates and upload them manually all at once (that will likely lead to some lost work if the area is frequently mapped by others, though) - preferably some time after they have been mapped and not immediately afterwards

Would it be conceivable to deactivate autoupload by default and display a privacy message when activating it?
And how about a function that automatically uploads edits only once a week?

**Comment by mnalis** [comment]
**TL;DR:** It wouldn't really help. As matkoniecz noted: uploading on-the-ground surveys to OSM to be publicly visible to whole world is simply incompatible with location privacy. 🤷‍♂️. 

If users (after reading SC [privacy policy](https://streetcomplete.app/privacy) and suggestions above) are still concerned about it, yet they still want to help the OSM project, they better stick with e.g. [HOT](https://wiki.openstreetmap.org/wiki/Humanitarian_OSM_Team) or [MapRoulette](https://wiki.openstreetmap.org/wiki/MapRoulette) (hey they have gamification too!) and map only in countries they've never been to, or contribute in other ways (translations, code, issue triage, donations, advertising for OSM etc).


---

First, let me prefece this by saying I'm definitely for increasing user privacy in any sensible way which does not have too much negative consequences. (Un)fortunately, SC is pretty much at that sweet spot already -- where adding even little more privacy-by-default is only possible with significant increase in inconveniences.

> Would it be conceivable to deactivate autoupload by default and display a privacy message when activating it?

Conceivable, yes (I have a great imagination! 😃)  Likely - I'd guess probably no. The StreetComplete target audience are not well-versed OSM mappers, most would not even know that they need to "upload" the answers. Sure, one could make tutorial bigger, and hope they read it and understand it (and remember it!), but still, it is likely very desirable to keep auto-upload enabled by default.

I'm afraid disabling auto-upload by default would bring a lot of bad experiences and confusions, less contributions, and for very little, if any, gain (by itself).

>  And how about a function that automatically uploads edits only once a week?

Meh... By itself, it wouldn't accomplish much. E.g. many (most?) people go to work and to home each day of the workweek, so gains of uploading once-a-week vs. once-a-day (e.g. if you set `upload on wifi only`) are very little.

Note that even when you implement **_all_** of my suggestions from that post above, you only partially (although noticeably) increase privacy. Someone targeting you will likely still be able to determine a lot of location intel even with all those precautions.  And even suggesting to SC users  to implement half of those protective measures is highly unrealistic IMHO (especially given target audience). 

Just ask yourself: how many of StreetComplete users have installed privacy-friendly custom ROM on their phone (e.g. [CalyxOS](https://calyxos.org/))? Or at very least de-googlified their phone (i.e. do not have Google account, don't use Google Play app store, Chromium, Google search engine, Youtube app, Google Mobile Services [e.g. "push" support] etc.) ? Because using _any_ of those services is realistically much bigger privacy leak than using StreetComplete. 

SC doesn't collect telemetry so we can't know, but I'd be willing to wager that number of people who even cared enough to **read** the (short and succinct and very readable!) in-app SC privacy policy before starting to use it is not much more than 1% (yes I know it is devastating). And those willing to also both remember and implement even most basic privacy-preserving [OpSec](https://en.wikipedia.org/wiki/Operations_security) suggestions I've mentioned above (much less something more advanced!) is likely noticeably below 0.1% of the userbase.
